2_process/src/process_nhdv2_attr.R

```diff
@@ -269,3 +269,184 @@ process_catchment_nhdv2_attr <- function(file_path,vars_table,segs_w_comids,nhd_
+    #all NHD reaches for this PRMS segment
+    nhd_reaches <- filter(prms_nhdv2_xwalk,
+                          PRMS_segid %in% ind_areas$PRMS_segid[i]) %>%
+      select(comid_cat) %>%
+      str_split(., pattern = ';', simplify = T)
+    
+    ind_areas$length_m[i] <- filter(nhdv2_reaches,
+                                    COMID %in% nhd_reaches) %>%
+      select(LENGTHKM) %>% st_drop_geometry() %>%
+      sum()
+  }
+  #Compute the reach stream density length (km)/area (sq.km)
+  #There must be a typo in the table's units because using m length gives
+  #results that are 3 orders of magnitude larger than other values
+  ind_areas <- mutate(ind_areas, str_dens = length_m/CAT_BASIN_AREA_sum) %>%
+    select(-length_m, -CAT_BASIN_AREA_sum)
+  #assign to attribute table
+  nhdv2_attr_refined <- mutate(nhdv2_attr_refined,
+                          CAT_STRM_DENS_area_wtd = case_when(PRMS_segid %in% ind_areas$PRMS_segid ~
+                                                               ind_areas$str_dens[match(PRMS_segid, ind_areas$PRMS_segid)],
+                                                             TRUE ~ CAT_STRM_DENS_area_wtd)
+  )
+  
+  #Compute TOT variables from PRMS CAT variables
+  # CWD, TAV7100, TMIN7100, STRM_DENS
+  #Add hru segment identifier to match with the recursive function output ID
+  nhdv2_attr_refined$hru_segment <- apply(str_split(nhdv2_attr_refined$PRMS_segid, 
+                                                    pattern = '_', simplify = T), 
+                                          MARGIN = 1, FUN = first) %>%
+    as.numeric()
+  
+  nhdv2_attr_refined$TOT_CWD_frmCAT <- 0
+  nhdv2_attr_refined$TOT_TAV7100_ANN_frmCAT <- 0
+  nhdv2_attr_refined$TOT_TMIN7100_frmCAT <- 0
+  nhdv2_attr_refined$TOT_STRM_DENS_frmCAT <- 0
+  #loop over all segments
+  for(i in 1:length(unique(nhdv2_attr_refined$hru_segment))){
+    #current segment
+    subseg <- unique(nhdv2_attr_refined$hru_segment)[i]
+    #all upstream segments, including current segment
+    ind_segs <- filter(prms_attribute_df, subseg_seg == subseg) %>% 
+      pull(all_from_segs)
+    #area for each PRMS segment
+    areas_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_BASIN_AREA_sum)
+    #CAT variables for each segment
+    CWD_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_CWD_area_wtd)
+    TAV_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_TAV7100_ANN_area_wtd)
+    TMN_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_TMIN7100_area_wtd)
+    STR_DENS_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_STRM_DENS_area_wtd)
+    
+    #Compute TOT values
+    #need special handling for PRMS segments with _2
+    ind_replace <- which(nhdv2_attr_refined$hru_segment == subseg)
+    nhdv2_attr_refined$TOT_CWD_frmCAT[ind_replace] <- compute_tot(CWD_segs,areas_segs)
```

Comment:
Your `compute_tot()` function does the trick, but {stats} included in base R also has a function called `weighted.mean` where you specify the weights using an argument `w`.